File: backend/src/test/java/com/worklog/application/service/DailyAttendanceServiceTest.java
package com.worklog.application.service;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import com.worklog.domain.attendance.DailyAttendance;
import com.worklog.domain.member.MemberId;
import com.worklog.domain.shared.DomainException;
import com.worklog.domain.tenant.TenantId;
import com.worklog.infrastructure.repository.JdbcDailyAttendanceRepository;
import java.time.LocalDate;
import java.time.LocalTime;
import java.util.Optional;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
@DisplayName("DailyAttendanceService")
class DailyAttendanceServiceTest {

    @Mock
    private JdbcDailyAttendanceRepository repository;

    @InjectMocks
    private DailyAttendanceService service;

    private static final TenantId TENANT = TenantId.of("550e8400-e29b-41d4-a716-446655440001");
    private static final MemberId MEMBER = MemberId.of("660e8400-e29b-41d4-a716-446655440001");

    @Test
    @DisplayName("saveAttendance should create new record when none exists")
    void saveAttendance_createsNew() {
        var date = LocalDate.of(2026, 3, 2);
        when(repository.findByMemberAndDate(MEMBER, date)).thenReturn(Optional.empty());

        service.saveAttendance(TENANT, MEMBER, date, LocalTime.of(9, 0), LocalTime.of(18, 0), "Test", 0);

        verify(repository).save(any(DailyAttendance.class));
    }

    @Test
    @DisplayName("saveAttendance should update existing with matching version")
    void saveAttendance_updatesExisting() {
        var date = LocalDate.of(2026, 3, 2);
        var existing = DailyAttendance.create(TENANT, MEMBER, date);
        when(repository.findByMemberAndDate(MEMBER, date)).thenReturn(Optional.of(existing));

        service.saveAttendance(TENANT, MEMBER, date, LocalTime.of(10, 0), LocalTime.of(19, 0), null, 0);

        verify(repository).save(existing);
        assertEquals(LocalTime.of(10, 0), existing.getStartTime());
    }

    @Test
    @DisplayName("saveAttendance should throw on version mismatch")
    void saveAttendance_throwsOnVersionMismatch() {
        var date = LocalDate.of(2026, 3, 2);
        var existing = DailyAttendance.create(TENANT, MEMBER, date);
        when(repository.findByMemberAndDate(MEMBER, date)).thenReturn(Optional.of(existing));

        // existing.version is 0, but we pass version 5
        assertThrows(
                DomainException.class,
                () -> service.saveAttendance(TENANT, MEMBER, date, LocalTime.of(9, 0), LocalTime.of(18, 0), null, 5));
    }

    @Test
    @DisplayName("deleteAttendance should delegate to repository")
    void deleteAttendance_delegates() {
        var date = LocalDate.of(2026, 3, 2);
        service.deleteAttendance(MEMBER, date);
        verify(repository).deleteByMemberAndDate(MEMBER, date);
    }
}

File: backend/src/test/kotlin/com/worklog/application/service/DailyAttendanceServiceTest.kt
package com.worklog.application.service

import com.worklog.domain.attendance.DailyAttendance
import com.worklog.domain.attendance.DailyAttendanceId
import com.worklog.domain.member.MemberId
import com.worklog.domain.tenant.TenantId
import com.worklog.infrastructure.repository.JdbcDailyAttendanceRepository
import io.mockk.Runs
import io.mockk.clearMocks
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDate
import java.time.LocalTime
import java.util.UUID

class DailyAttendanceServiceTest {

    private val attendanceRepository = mockk<JdbcDailyAttendanceRepository>()
    private lateinit var service: DailyAttendanceService

    private val testTenantId = TenantId.of(UUID.randomUUID())
    private val testMemberId = MemberId.of(UUID.randomUUID())
    private val testDate = LocalDate.of(2026, 3, 1)
    private val testStartTime = LocalTime.of(9, 0)
    private val testEndTime = LocalTime.of(18, 0)

    @BeforeEach
    fun setup() {
        clearMocks(attendanceRepository)
        service = DailyAttendanceService(attendanceRepository)
    }

    @Test
    fun `saveAttendance should create new when none exists`() {
        every { attendanceRepository.findByMemberAndDate(testMemberId, testDate) } returns null
        every { attendanceRepository.save(any()) } just Runs

        val result = service.saveAttendance(
            testTenantId,
            testMemberId,
            testDate,
            testStartTime,
            testEndTime,
            "Test remarks",
            null,
        )

        assertNotNull(result)
        verify(exactly = 1) { attendanceRepository.save(any()) }
    }

    @Test
    fun `saveAttendance should update existing with matching version`() {
        val existingId = DailyAttendanceId.generate()
        val existing = DailyAttendance(
            existingId,
            testTenantId,
            testMemberId,
            testDate,
            LocalTime.of(8, 0),
            LocalTime.of(17, 0),
            "Old remarks",
            1,
        )

        every { attendanceRepository.findByMemberAndDate(testMemberId, testDate) } returns existing
        every { attendanceRepository.save(any()) } just Runs

        val result = service.saveAttendance(
            testTenantId,
            testMemberId,
            testDate,
            testStartTime,
            testEndTime,
            "Updated remarks",
            1,
        )

        assertEquals(existingId.value(), result)
        verify(exactly = 1) { attendanceRepository.save(any()) }
    }

    @Test
    fun `saveAttendance should throw on version mismatch`() {
        val existing = DailyAttendance(
            DailyAttendanceId.generate(),
            testTenantId,
            testMemberId,
            testDate,
            LocalTime.of(8, 0),
            LocalTime.of(17, 0),
            "Old remarks",
            2,
        )

        every { attendanceRepository.findByMemberAndDate(testMemberId, testDate) } returns existing

        assertThrows(IllegalStateException::class.java) {
            service.saveAttendance(
                testTenantId,
                testMemberId,
                testDate,
                testStartTime,
                testEndTime,
                "Updated remarks",
                1,
            )
        }

        verify(exactly = 0) { attendanceRepository.save(any()) }
    }

    @Test
    fun `deleteAttendance should delegate to repository`() {
        every { attendanceRepository.deleteByMemberAndDate(testMemberId, testDate) } just Runs

        service.deleteAttendance(testMemberId, testDate)

        verify(exactly = 1) { attendanceRepository.deleteByMemberAndDate(testMemberId, testDate) }
    }
}
